src/bank.rs

```diff
@@ -1228,18 +1351,22 @@ impl Bank {
         self.last_ids
             .read()
             .unwrap()
-            .sigs
+            .entries
             .get(last_id)
-            .and_then(|sigs| sigs.0.get(signature).cloned())
+            .and_then(|entry| entry.signature_status.get(signature).cloned())
     }
 
     /// Hash the `accounts` HashMap. This represents a validator's interpretation
-    ///  of the ledger up to the `last_id`, to be sent back to the leader when voting.
+    ///  of the delta of the ledger since the last vote and up to now
     pub fn hash_internal_state(&self) -> Hash {
         let mut ordered_accounts = BTreeMap::new();
-        for (pubkey, account) in self.accounts.read().unwrap().iter() {
+
+        // only hash internal state of the part being voted upon, i.e. since last
```

Comment:
@garious @aeyakovenko is this bit of code is correct?

It implies that we expect all nodes to implement this voting hash functionality the same way, i.e. that all nodes voting on this segment of the ledger have **exactly the same checkpoint delta**.  I think it's right, but would appreciate a sanity check.

Reply:
Correct that the algorithm to generate the voting hash is part of the network protocol. Please make sure this particular flavor makes it into an RFC sooner than later. Once there, I'll add it to the architecture doc.